cachi2/core/package_managers/rubygems/main.py

```diff
@@ -0,0 +1,95 @@
+import logging
+from typing import Optional
+
+from cachi2.core.models.input import Request
+from cachi2.core.models.output import EnvironmentVariable, ProjectFile, RequestOutput
+from cachi2.core.models.sbom import Component
+from cachi2.core.package_managers.rubygems.parser import (
+    BundlerDependency,
+    GemDependency,
+    GitDependency,
+    PathDependency,
+)
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def fetch_rubygems_source(request: Request) -> RequestOutput:
+    """Resolve and process all rubygems packages."""
+    components: list[Component] = []
+    environment_variables: list[EnvironmentVariable] = []
+    project_files: list[ProjectFile] = []
+
+    for package in request.packages:
+        package_dir = request.source_dir.join_within_root(package.path)
+        _resolve_rubygems(package_dir=package_dir, output_dir=request.output_dir)
+
+    return RequestOutput.from_obj_list(
+        components=components,
+        environment_variables=environment_variables,
+        project_files=project_files,
+    )
+
+
+def _resolve_rubygems(package_dir: RootedPath, output_dir: RootedPath) -> list[Component]:
+    """
+    Process a request for a single rubygems package.
+
+    TODO:
+    1. Parse Gemfile.lock
+    2. Get project metadata
+    3. Download dependencies
+    4. Generate SBOM components
+    """
+    return NotImplemented
+
+
+def _get_package_metadata(
+    package_dir: RootedPath,
+    path_dependencies: list[PathDependency],
+) -> tuple[str, Optional[str]]:
+    """TODO."""
+    return NotImplemented
+
+
+def _get_repository_name(package_dir: RootedPath) -> str:
+    """TODO."""
+    return NotImplemented
+
+
+def _download_dependencies(output_dir: RootedPath, dependencies: list[BundlerDependency]) -> None:
+    """TODO."""
+
+
+def _download_rubygems_package(output_dir: RootedPath, gem: GemDependency) -> None:
+    """TODO."""
+
+
+def _download_git_package(output_dir: RootedPath, gem: GitDependency) -> None:
+    """
+    TODO.
+
+    # directory has a specific format: {repository_name}-{revision}
+    # revision must have 12 characters
+    # https://github.com/rubygems/rubygems/blob/3da9b1dda0824d1d770780352bb1d3f287cb2df5/bundler/lib/bundler/source/git.rb#L327
+    """
+
+
+def _bundle_config_exists(package_dir: RootedPath) -> bool:
+    """TODO."""
+    return NotImplemented
+
+
+def _inject_bundle_config(package_dir: RootedPath) -> None:
+    """TODO."""
+
+
+def _generate_environment_variables() -> list[EnvironmentVariable]:
+    """TODO."""
+    return NotImplemented
+
+
+def _generate_purl_for(gem: BundlerDependency) -> str:
```

Comment:
A nitpicky question: I noticed that purls are often typed as strings when in reality they are obviously purls. Is there a reason for not having `Purl` class subclassing `str`? This question holds for other fields in types defined in https://github.com/containerbuildsystem/cachi2/pull/584/files#diff-7c68c39848a484d1335d5268342184ecf184fc0a740e1203e376230d7a35cd62

Reply:
Will move this discussion to some other place, resolving here to not hold the PR.

Reply:
I think it might be a good idea to have a centralized PURL class, but then, I think we should have a separate PR that applies it project-wide, not just on the Ruby part.

Reply:
Yes, this is definitely out of scope for this PR.